### PR TITLE
Order and OrderItem models hw

### PR DESCRIPTION
**Pull request by RostyslavKuzmych**

### CODE REVIEW

File: src/main/java/application/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    @PatchMapping("/{id}")
+    @ResponseStatus(HttpStatus.OK)
+    @Operation(summary = "Update order status",
+            description = "An endpoint for updating an order status")
+    @PreAuthorize("hasRole('ADMIN')")
+    public void updateStatusOrder(@PathVariable Long id,
+                                  @RequestBody OrderRequestStatusDto dto) {
+        orderService.updateStatusOrder(id, dto);
```
Comment (vovasalyha) [inline src/main/java/application/controller/OrderController.java:64]: I think it's better to name `updateOrderStatus`

File: src/main/java/application/exception/CustomGlobalExceptionHandler.java [hunk 2]
```diff
@@ -45,6 +45,24 @@ protected ResponseEntity<Object> handleRegistrationException(RegistrationExcepti
         return new ResponseEntity<>(body, HttpStatus.BAD_REQUEST);
     }
 
+    @ExceptionHandler(EntityNotFoundException.class)
+    protected ResponseEntity<Object> handleEntityNotFoundException(EntityNotFoundException ex) {
+        Map<String, Object> body = new LinkedHashMap<>();
```
Comment (vovasalyha) [inline src/main/java/application/exception/CustomGlobalExceptionHandler.java:50]: Don't duplicate the same logic multiple times. Remember about KISS

File: src/main/java/application/model/Order.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+@SQLDelete(sql = "UPDATE orders SET is_deleted = true WHERE id = ?")
+@Where(clause = "is_deleted = false")
+@Entity
+@Table(name = "orders")
+public class Order {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne
```
Comment (vovasalyha) [inline src/main/java/application/model/Order.java:37]: Specify mappings properly. Also, remind to yourself, why is it bad to use eager relations?

File: src/main/java/application/model/Order.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    @NonNull
+    private User user;
+
+    @Column(nullable = false)
+    @Enumerated(EnumType.STRING)
+    private Status status;
+
+    @NonNull
```
Comment (vovasalyha) [inline src/main/java/application/model/Order.java:45]: I'm not sure if it's necessary to use this annotation in entities.

File: src/main/java/application/model/Order.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    private BigDecimal total = BigDecimal.ZERO;
+
+    @NonNull
+    private LocalDateTime localDateTime;
```
Comment (vovasalyha) [inline src/main/java/application/model/Order.java:49]: Create a better name for this variable (see task requirements)

File: src/main/java/application/model/OrderItem.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+@Table(name = "order_items")
+public class OrderItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne
+    @JoinColumn(name = "order_id")
+    @NonNull
+    private Order order;
```
Comment (vovasalyha) [inline src/main/java/application/model/OrderItem.java:31]: Don't forget to exclude relations from equals and hash code

File: src/main/java/application/service/OrderItemService.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
+package application.service;
+
+import application.dto.item.OrderItemResponseDto;
+import application.model.OrderItem;
+import java.util.List;
+import org.springframework.security.core.Authentication;
+
+public interface OrderItemService {
+    OrderItem save(OrderItem orderItem);
```
Comment (vovasalyha) [inline src/main/java/application/service/OrderItemService.java:9]: Be consistent in your services regarding returning of entities or DTOs. If you use DTOs, use them for all services, but don't mix entity/DTO in a service.

File: src/main/java/application/service/OrderService.java [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+public interface OrderService {
+    OrderResponseDto createOrder(ShoppingCart shoppingCart,
+                                 OrderRequestShippingAddressDto dto);
+
+    List<OrderResponseDto> getListOrderResponseDtos(Long id);
+
+    void updateStatusOrder(Long id, OrderRequestStatusDto dto);
+
+    List<Order> findAllByUserId(Long id);
```
Comment (vovasalyha) [inline src/main/java/application/service/OrderService.java:18]: The same here (check all your services methods and ensure you are consistent in this choice)

File: src/main/java/application/service/impl/OrderServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+        OrderResponseDto responseDto = orderMapper.toResponseDto(order);
+        return setOrderItemsDtos(responseDto, order);
+    }
+
+    @Override
+    public List<OrderResponseDto> getListOrderResponseDtos(Long id) {
+        return findAllByUserId(id)
+                .stream()
+                .map(orderMapper::toResponseDto)
+                .map(c -> setOrderItemsDtos(c, findByOrderId(c.getId())
```
Comment (vovasalyha) [inline src/main/java/application/service/impl/OrderServiceImpl.java:54]: Is it possible to perform this mapping in the mapper?

File: src/main/java/application/service/impl/OrderServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    @Override
+    public Order findByOrderId(Long id) {
+        return orderRepository.findById(id)
+                .orElseThrow(() ->
+                        new EntityNotFoundException(FIND_ORDER_EXCEPTION + id));
+    }
+
+    private Order setAllFieldsOrder(ShoppingCart sc, OrderRequestShippingAddressDto dto) {
```
Comment (vovasalyha) [inline src/main/java/application/service/impl/OrderServiceImpl.java:78]: Please analyse all your method names on grammar mistakes. In this context, more correct would be `setAllOrderFields`, but it's also not the best name for this method. Think about something like `prepareOrder` or similar.

File: src/main/java/application/service/impl/OrderServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    private BigDecimal getTotalPrice(Set<OrderItem> orderItems) {
+        BigDecimal bigDecimal = orderItems.stream()
+                .map(OrderItem::getPrice)
+                .reduce(BigDecimal.ZERO, BigDecimal::add);
+        return bigDecimal;
+    }
+
+    private Set<OrderItem> saveSetOrderItemsFromShoppingCart(Order order,
```
Comment (vovasalyha) [inline src/main/java/application/service/impl/OrderServiceImpl.java:112]: Also a bit strange method name.

File: src/main/resources/db/changelog/changes/16-create-orders-table.yaml [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+                  name: shipping_address
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
+              - column:
+                  name: is_deleted
+                  type: boolean
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
```
Comment (vovasalyha) [inline src/main/resources/db/changelog/changes/16-create-orders-table.yaml:49]: Don't forget about empty lines.

**Review by vovasalyha — CHANGES_REQUESTED** [review]
See a couple of comments.

File: src/main/java/application/model/Order.java (hunk 3 above)
Comment (kshuryhin) [reply src/main/java/application/model/Order.java:37]: @RostyslavKuzmych do you think here will be `EAGER` or `LAZY` by default?

File: src/main/java/application/model/OrderItem.java (hunk 6 above)
Comment (kshuryhin) [reply src/main/java/application/model/OrderItem.java:30]: I guess still a valid point to consider

File: src/main/java/application/service/OrderItemService.java (hunk 7 above)
Comment (kshuryhin) [reply src/main/java/application/service/OrderItemService.java:9]: Still valid point

File: src/main/java/application/service/impl/OrderServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    private BigDecimal getTotalPrice(Set<OrderItem> orderItems) {
+        BigDecimal bigDecimal = orderItems.stream()
+                .map(OrderItem::getPrice)
+                .reduce(BigDecimal.ZERO, BigDecimal::add);
+        return bigDecimal;
+    }
+
+    private Set<OrderItem> setOrderItemsToOrderFromShoppingCart(Order order,
```
Comment (kshuryhin) [inline src/main/java/application/service/impl/OrderServiceImpl.java:114]:
```suggestion
    private Set<OrderItem> populateWithOrderItems(Order order,

```
I guess that could be something like that

File: src/main/java/application/dto/order/OrderRequestStatusDto.java [hunk 14]
```diff
@@ -0,0 +1,9 @@
+package application.dto.order;
+
+import application.model.Order;
+import lombok.Data;
+
+@Data
+public class OrderRequestStatusDto {
+    private Order.Status status;
```
Comment (kshuryhin) [inline src/main/java/application/dto/order/OrderRequestStatusDto.java:8]: Just wondering if you use this enum in several places outside of the `Order` class, probably that would be better to move it to the separate file?

File: src/main/java/application/model/OrderItem.java [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+import jakarta.persistence.Table;
+import java.math.BigDecimal;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Entity
+@Data
+@NoArgsConstructor
```
Comment (kshuryhin) [inline src/main/java/application/model/OrderItem.java:19]: Do we need this?

File: src/main/java/application/service/OrderItemService.java [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+
+import application.dto.item.OrderItemResponseDto;
+import application.model.OrderItem;
+import java.util.List;
+import org.springframework.security.core.Authentication;
+
+public interface OrderItemService {
+    OrderItem save(OrderItem orderItem);
+
+    List<OrderItemResponseDto> getAllOrderItemsDtoByOrderId(Authentication authentication,
```
Comment (kshuryhin) [inline src/main/java/application/service/OrderItemService.java:11]: Actually not sure if is is useful to mention `authentication` in the service contract. First of all, we can call it from the corresponding service and not pass as an argument. But I guess the most preferable here would be to use some method level `@PreAuthorize` with custom permission check
https://docs.spring.io/spring-security/reference/servlet/authorization/method-security.html

**Review by kshuryhin — CHANGES_REQUESTED** [review]
Good job, but there are still some points to consider

**Pull request merged**